### ISSUE TITLE
Implementing "client_assertion" 

### ISSUE THREAD

**Issue by naizerjohn-ms** [golang/oauth2#744]
Hello!

Currently this package does not support client_assertion/client_assertion_type OAuth2.0 client authentication outlined [here](https://openid.net/specs/openid-connect-core-1_0.html#ClientAuthentication) in the OpenID Connect documentation (not up to standard). Here is an example outlined in this documentation, for a visual on what the request would look like:

```  POST /token HTTP/1.1
  Host: server.example.com
  Content-Type: application/x-www-form-urlencoded

  grant_type=authorization_code&
    code=i1WsRn1uB1&
    client_id=s6BhdRkqt3&
    client_assertion_type=urn%3Aietf%3Aparams%3Aoauth%3Aclient-assertion-type%3Ajwt-bearer&
    client_assertion=PHNhbWxwOl ... ZT
```

as opposed to what this package only currently supports (client_secret):

```  POST /token HTTP/1.1
  Host: server.example.com
  Content-Type: application/x-www-form-urlencoded

  grant_type=authorization_code&
    code=i1WsRn1uB1&
    client_id=s6BhdRkqt3&
    client_secret=PHNhbWxwOl ... ZT
```

I am willing to work towards this implementation and am asking for any support/guidance for achieving this solution. Many tech companies (including ours) are migrating away from using secrets and towards more secure authentication systems. Please see [this](https://blog.logto.io/client-assertion-in-client-authn/) article which provides a deeper description on what client_assertions are as well.

**Comment by zetaab** [comment]
we are interested of this feature as well.

**Comment by zetaab** [comment]
@naizerjohn-ms that can be done already?

Example (disclaimer: I did not test this but afaik when reading code it should be possible)
```
	conf := &oauth2.Config{
		ClientID:     authConf.ClientID,
		Endpoint:     provider.Endpoint(),
		Scopes:       scopes,
		RedirectURL:  authConf.RedirectURI,
		AuthStyle:    oauth2.AuthStyleInParams,
	}
	oauthCtx := oidc.ClientContext(context.Background(), &http.Client{})
	otoken, err := conf.Exchange(
		oauthCtx,
		content.Code,
		oauth2.SetAuthURLParam("client_assertion", "foo"),
		oauth2.SetAuthURLParam("client_assertion_type", "urn:ietf:params:oauth:client-assertion-type:jwt-bearer"),
	)
	...
```

**Comment by naizerjohn-ms** [comment]
@zetaab Great point, let me implement this and test... looks like it should work? I'm looking at oauth2/internal/token.go at function 'newTokenRequest', and it seems it only adds the client_secret to the body of the request if it is not null/empty

**Comment by zetaab** [comment]
@naizerjohn-ms yep it adds it only if its defined, but you can define extra parameters with oauth2.SetAuthURLParam before that. Please test it if you have use-case :)

**Comment by francisconunesnavarro** [comment]
we are also interested in this feature as well, waiting for the PR to be approve

**Comment by naizerjohn-ms** [comment]
@zetaab This has been tested and verified. Will close issue now!

**Comment by naizerjohn-ms** [comment]
Hello @zetaab! I am reopening this issue, but I may just create a new one for specificity. Your recommendation above as stated has been verified to work which is great! There is one issue though... the token refresh does not work when using the method above by setting the authURLparams. This is because when this package does a token refresh it calls the TokenSource method inside oauth2.go file here:
![image](https://github.com/user-attachments/assets/538c433f-1063-44ea-b36f-e8bae1e74b9f)

TokenSource does NOT support passing in additional authURLparams like the Exchange method does. Any help to get this support implemented would be much appreciated! Or if I missed something, please show me!

**Comment by zetaab** [comment]
@naizerjohn-ms yeah well might be.. but the main issue is that none is reviewing. I have no idea how to get reviews

**Comment by naizerjohn-ms** [comment]
Adding @bradfitz seeing that you have made contributions recently!

**Comment by naizerjohn-ms** [comment]
@cuishuang seeing that you have made a contribution recently, how may I be able to get someone to review/merge a PR?

**Comment by michaeldcanady** [comment]
I'd love to be able to implement this!

**Comment by zetaab** [comment]
@naizerjohn-ms & co. I got reply finally:

> This needs a https://go.dev/s/proposal

If someone do have time to continue with this feel free. I am not anymore working in job where I was working and this feature was needed.

**Comment by michaeldcanady** [comment]
Thank you! Created [Proposal #73431](https://github.com/golang/go/issues/73431)
